TYP: type ma.min

luxedo opened the pull request:
<!--         ----------------------------------------------------------------
                MAKE SURE YOUR PR GETS THE ATTENTION IT DESERVES!
                ----------------------------------------------------------------

*  FORMAT IT RIGHT:
      https://www.numpy.org/devdocs/dev/development_workflow.html#writing-the-commit-message

*  IF IT'S A NEW FEATURE OR API CHANGE, TEST THE WATERS:
      https://www.numpy.org/devdocs/dev/development_workflow.html#get-the-mailing-list-s-opinion

*  HIT ALL THE GUIDELINES:
      https://numpy.org/devdocs/dev/index.html#guidelines

*  WHAT TO DO IF WE HAVEN'T GOTTEN BACK TO YOU:
      https://www.numpy.org/devdocs/dev/development_workflow.html#getting-your-pr-reviewed
-->
Add static typing to `ma.min`. Part of #26404.

luxedo committed:
TYP: type ma.min

luxedo commented:
Any updates on this?

Jacob-Stevens-Haas commented on numpy/ma/core.pyi:
```diff
@@ -371,7 +374,13 @@ class _extrema_operation(_MaskedUFunc):
 
-def min(obj, axis=..., out=..., fill_value=..., keepdims=...): ...
+def min(
+    obj: MaskedArray[_ShapeLike, _SCT] | _ArrayLike[_SCT],
```
Do you need `_ArrayLike[_SCT]`?  Isn't this file specifically `MaskedArray` (asking b/c I'm not sure).

Jacob-Stevens-Haas commented on numpy/ma/core.pyi:
```diff
@@ -371,7 +374,13 @@ class _extrema_operation(_MaskedUFunc):
-def min(obj, axis=..., out=..., fill_value=..., keepdims=...): ...
+def min(
+    obj: MaskedArray[_ShapeLike, _SCT] | _ArrayLike[_SCT],
+    axis: None | _ShapeLike = ...,
+    out: None = ...,
+    fill_value: None | _ScalarLike_co = ...,
+    keepdims: None | bool = ...,
+) -> _SCT: ...
```
if `axis` is `_ShapeLike` and not `None`, the return type will be the same type as `obj`, not `_SCT`.

Jacob-Stevens-Haas reviewed (commented):
So I'm working on #26081 and while not a numpy dev, I can hopefully ask and answer some questions.

1. `np.amin` has three overloads to tell the type checker three different cases
    * with no axis argument, `amin` will return the argmin of the entire array as a python (not numpy) object, so tell the type checker that.
    * With an axis argument and no `out`, it's impossible to tell what `_ArrayLike` object `a` is, so tell type checker `Any`
    * With an argument `out`, we know that the returned variable will be `out`, so the type checker knows the return type matches `out`
These are probably also relevant for masked arrays.
2. I don't think you need more documentation, because there's no new types, other than an entry in doc/release/upcoming_changes/26624.improvement.rst.
3. Yes, most likely.  A file numpy/typing/tests/data/pass/masked.py and likewise .../fail/masked.py that will fill up with type checks as #26404 is implemented.  This is a good place to test-drive assumptions you might have about how your type annotations work as you develop the PR (e.g. the comment on `axis` and `obj` return types)

luxedo commented:
@Jacob-Stevens-Haas thank you for the feedback. I'll be working on this next week.

charris commented:
Need rebase.

luxedo committed:
TYP: add ma.main typing tests

luxedo commented:
Rebased and upated with @Jacob-Stevens-Haas hints. Also added passing typing tests.

I couldn't understand how to make failing tests. When calling `spin mypy` somehow it kept throwing me old errors even after deleting the test file.

Jacob-Stevens-Haas reviewed (commented):
The PR is great, a few comments about tests (with a grain of salt - I'm learning too):

> I couldn't understand how to make failing tests. 

Fail tests are written with arguments that fail type checking, with a comment afterwards for the expected type checking error (mostly either "no overload variant" or "argument has incompatible type" e.g. 

https://github.com/numpy/numpy/blob/d52816ae9f95781cc2a48cac0a8ca540e051f648/numpy/typing/tests/data/fail/fromnumeric.pyi#L126-L130

Edit: But thinking about it more, there's no illegal combination I can think of worth testing.  E.g. the overloads allow `np.min(1.0)` and `np.min(arr, axis=0)`.  If they prevented something like `np.ma.min(1.0, axis=0)`, that would be a good fail test.  I don't think there are any such holes in the overloads for this PR though, so fail tests may not be necessary

There's also reveal tests that verify the type that gets returned with `assert_type`.  My mental comparison is that pass/fail tests check arguments, reveal tests check return value.  In this case, it might be helpful to check the returned type is not `Any` in overloads 1 and 3.

> When calling spin mypy somehow it kept throwing me old errors even after deleting the test file.

Did you try running `$ git clean -xdf`?

luxedo committed:
TST: fixing ma.min typing tests

Co-Authored-By: Jakob Stevens Haas <37048747+Jacob-Stevens-Haas@users.noreply.github.com>

luxedo commented:
@Jacob-Stevens-Haas I updated the PR with your suggestions

Jacob-Stevens-Haas reviewed (changes requested):
Thanks for updating this!  Can you add a reveal test to check the return values?  I.e. numpy/typing/tests/data/reveal/ma.pyi like:



```python
import numpy as np

arr: np.ma.MaskedArray[Any, np.dtype[np.float64]] = np.ma.masked_array([[1.0, 2.0]], mask=[True, False])
result = np.ma.min(arr, axis=None)
assert_type(result, np.float64)

tgt: np.ma.MaskedArray[Any, np.dtype[np.float64]] = np.ma.empty(shape=())
result = np.ma.min(arr, axis=None, out=tgt)
assert_type(result, np.ma.MaskedArray)

```
This verifies that the result type changes with the `out` parameter.  Since reveal tests even need to be executable, so you can remove the assignment for readability.

luxedo committed:
TST: add ma.min reveal typing tests

Co-Authored-By: Jakob Stevens Haas <37048747+Jacob-Stevens-Haas@users.noreply.github.com>

luxedo commented:
Both my `numpy`'s and `scipy`'s build are broken since last week :( 

I pushed this test and it failed. I'll try to make `spin mypy` working and then I fix the typing.

Jacob-Stevens-Haas commented:
Hmm, this is curious.  `keepdims` defaults to `_NoValue`, not `None` or `False`, even though the default behavior is `False`.  Thus the static typing sees the failing test as overload 2, instead of overload 1. 

> If the default value is passed, then keepdims will not be passed through to the min method of sub-classes of [ndarray](https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html#numpy.ndarray), however any non-default value will be. If the sub-class’ method does not implement keepdims any exceptions will be raised.

In other words, `_NoValueType` exists in order to allow subclasses to break Liskov Substitution, which means might be a problem in and of itself.  It looks like `_NoValueType` [was first discussed](https://github.com/numpy/numpy/pull/5683#issuecomment-80823178) as a sort of third option for an argument and has come up again in a few PRs and issues [1], [2], and [3].

The long and short of this is that you can _probably_ include `_NoValueType` in overload 1, e.g. `keep_dims: Literal[False] | _NoValueType`.  But it would also be good to check whether we really need to allow `_NoValue` as the default or we can just default to false.

[1]: #8991 
[2]: #13105 
[3]: #19173

jorenham commented on numpy/typing/tests/data/reveal/ma.pyi:
```diff
@@ -0,0 +1,13 @@
+from typing_extensions import assert_type
+
+arr: np.ma.MaskedArray[Any, np.dtype[np.float64]] = np.ma.masked_array([[1.0, 2.0]], mask=[True, False])
+result = np.ma.min(arr, axis=None)
+assert_type(result, np.float64)
```
by inlining the `result`, you avoid changing the type `result` later on, which technically isn't valid (mypy allowing this is a bug IMHO).

```suggestion
assert_type(np.ma.min(arr, axis=None), np.float64)
```

jorenham commented on numpy/typing/tests/data/reveal/ma.pyi:
```diff
@@ -0,0 +1,13 @@
+result = np.ma.min(arr, axis=None)
+assert_type(result, np.float64)
+
+tgt: np.ma.MaskedArray[Any, np.dtype[np.float64]] = np.ma.empty(shape=())
+result = np.ma.min(arr, axis=None, out=tgt)
+assert_type(result, np.ma.MaskedArray)
```
The type of `tgt` is equivalent to that of `arr`, so there's no need for it (so it doesn't matter that at runtime this wouldn't make a lot of sense).

Additionally, `np.ma.MaskedArray` should have type arguments, as its type parameters have no defaults (PEP 696).

```suggestion
assert_type(np.ma.min(arr, axis=None, out=arr), np.ma.MaskedArray[Any, np.dtype[np.float64]])
```

jorenham commented on numpy/ma/core.pyi:
```diff
@@ -14,9 +14,12 @@ from numpy import (
     zeros_like as zeros_like,
     angle as angle
 )
+from numpy._typing import _ArrayLike, _ScalarLike_co, _ShapeLike, overload
```
It's probably safer to import `overload` directly from `typing` instead.

jorenham commented on numpy/ma/core.pyi:
```diff
@@ -14,9 +14,12 @@ from numpy import (
 
 _ShapeType_co = TypeVar("_ShapeType_co", bound=tuple[int, ...], covariant=True)
 _DType_co = TypeVar("_DType_co", bound=dtype[Any], covariant=True)
+_ArrayType = TypeVar("_ArrayType", bound="MaskedArray[Any, Any]")
```
There's no need for stringified annotations within `.pyi` stubs (and disallowed by e.g. `flake8-pyi` and several type-checkers)

jorenham reviewed (changes requested):
Thanks for this!

I left a couple of minor suggestions, but in general, I like the way this looks 🙁.

jorenham commented:
I see that there's also a (tiny) merge conflict.

jorenham commented:
This has been superseded by #28593, so I'm going to close this.
